platooning_tactical_plugin/src/platooning_tactical_plugin.cpp

```diff
@@ -0,0 +1,617 @@
+#include <algorithm>
+#include <memory>
+#include <boost/uuid/uuid_generators.hpp>
+#include <boost/uuid/uuid_io.hpp>
+#include <lanelet2_core/geometry/Point.h>
+#include <trajectory_utils/trajectory_utils.h>
+#include <trajectory_utils/conversions/conversions.h>
+#include <sstream>
+#include <carma_utils/containers/containers.h>
+#include <Eigen/Core>
+#include <Eigen/Geometry>
+#include <Eigen/LU>
+#include <Eigen/SVD>
+#include <platooning_tactical_plugin/smoothing/SplineI.h>
+#include <platooning_tactical_plugin/smoothing/CubicSpline.h>
+#include <platooning_tactical_plugin/log/log.h>
+#include <carma_utils/containers/containers.h>
+#include <platooning_tactical_plugin/smoothing/filters.h>
+#include <unordered_set>
+
+using oss = std::ostringstream;
+
+namespace platooning_tactical_plugin
+{
+PlatooningTacticalPlugin::PlatooningTacticalPlugin(carma_wm::WorldModelConstPtr wm, PlatooningTacticalPluginConfig config,
+                                           PublishPluginDiscoveryCB plugin_discovery_publisher)
+  : wm_(wm), config_(config), plugin_discovery_publisher_(plugin_discovery_publisher)
+{
+  plugin_discovery_msg_.name = "PlatooningTacticalPlugin";
+  plugin_discovery_msg_.versionId = "v1.0";
+  plugin_discovery_msg_.available = true;
+  plugin_discovery_msg_.activated = false;
+  plugin_discovery_msg_.type = cav_msgs::Plugin::TACTICAL;
+  plugin_discovery_msg_.capability = "tactical_plan/plan_trajectory";
+}
+
+bool PlatooningTacticalPlugin::onSpin()
+{
+  plugin_discovery_publisher_(plugin_discovery_msg_);
+  return true;
+}
+
+bool PlatooningTacticalPlugin::plan_trajectory_cb(cav_srvs::PlanTrajectoryRequest& req,
+                                              cav_srvs::PlanTrajectoryResponse& resp)
+{
+  ros::WallTime start_time = ros::WallTime::now(); // Start timeing the execution time for planning so it can be logged
+
+  lanelet::BasicPoint2d veh_pos(req.vehicle_state.X_pos_global, req.vehicle_state.Y_pos_global);
+  double current_downtrack = wm_->routeTrackPos(veh_pos).downtrack;
+
+  auto points_and_target_speeds = maneuvers_to_points(req.maneuver_plan.maneuvers, current_downtrack, wm_); // Convert maneuvers to points
+
+  ROS_DEBUG_STREAM("points_and_target_speeds: " << points_and_target_speeds.size());
+
+  auto downsampled_points =
+      carma_utils::containers::downsample_vector(points_and_target_speeds, config_.downsample_ratio);
+
+  ROS_DEBUG_STREAM("downsample_points: " << downsampled_points.size());
+
+  ROS_DEBUG_STREAM("PlanTrajectory");
+
+  cav_msgs::TrajectoryPlan trajectory;
+  trajectory.header.frame_id = "map";
+  trajectory.header.stamp = ros::Time::now();
+  trajectory.trajectory_id = boost::uuids::to_string(boost::uuids::random_generator()());
+
+  trajectory.trajectory_points = compose_trajectory_from_centerline(downsampled_points, req.vehicle_state); // Compute the trajectory
+  trajectory.initial_longitudinal_velocity = std::max(req.vehicle_state.longitudinal_vel, config_.minimum_speed);
```

Comment:
I believe this is also the actual vehicle speed at the start time of the trajectory, so likely just current vehicle speed if that data is available to this plugin.

Reply:
fixed